### ISSUE TITLE
Missing osiSockAttach() call leads to hang on exit

### ISSUE THREAD

**Issue by mdavidsaver** [epics-base/pvAccessCPP#81]
@MarkRivers report the following on Windows.

```
epics> var atExitDebug 1
epics> exit
atExit exitDatabase(0000000000000000)
atExit ca_client_exit_handler(0000000000000000)
atExit cacExitHandler(0000000000000000)
atExit twdShutdown(0000000000000000)
atExit syncShutdown(0000000000000000)
atExit dbndShutdown(0000000000000000)
atExit arrShutdown(0000000000000000)
atExit errlogExitHandler(0000000000000000)
atExit &pva_factory_cleanup(0000000000000000)
epicsSocketDestroy: failed to close a socket because "WINSOCK Error 10093"
2017-12-27T22:48:54.048 Receive thread for UDP socket 0.0.0.0:0 has not exited.
epicsSocketDestroy: failed to close a socket because "WINSOCK Error 10093"
```

**Comment by mdavidsaver** [comment]
Winsock error 10093 is WSANOTINITIALISED https://msdn.microsoft.com/en-us/library/windows/desktop/ms740668(v=vs.85).aspx

```
Successful WSAStartup not yet performed.  Either the application has not called  WSAStartup or  WSAStartup failed. The application may be accessing a  socket that the current active task does not own (that is, trying to  share a socket between tasks), or  WSACleanup has been called too many times.
```

I think we can rule out "WSAStartup failed" as osiSockAttach() would print an error in this case.  As socket creation was previously successful we can rule out WSAStartup not being called.  This leaves only WSACleanup being called too early.

**Comment by mdavidsaver** [comment]
As a note.  osiSockAttach()/Detach() are using a global counter without locking.  This should be switched to an atomic counter.

**Comment by mdavidsaver** [comment]
The server context is being shutdown from a global destructor, so after epicsExit() has run all handlers.

https://github.com/epics-base/pvAccessCPP/blob/c4fe7150c96ff7c391e41d61c28ebf06c7a1f151/src/ioc/PVAServerRegister.cpp#L43

**Comment by mdavidsaver** [comment]
Candidate fix d1fca67cf297f78aba15eaa0d9b511097c416710.  This moves PVA server shutdown into the epicsExit() sequence, and places it earlier than the places where osiSockRelease() is called.  Doesn't address the fundamental issue of where else osiSockAttach() should be called.  osiSockAttach() is already being called in what seem to me like reasonable places.

**Comment by mdavidsaver** [comment]
@MarkRivers reports that the error message is gone.  The larger issue remains, but I don't intend to pursue it further at this point.  So this will be a "wontfix" until/unless someone else wants to look into it.

**Comment by mdavidsaver** [comment]
Released with Base 7.0.2
